npm/podController.go

```diff
@@ -154,154 +163,353 @@ func appendNamedPortIpsets(ipsMgr *ipsm.IpsetManager, portList []v1.ContainerPor
 		}
 
 		label := podLabelKey + ":" + podLabelVal
 		log.Logf("Adding pod %s to ipset %s", podIP, label)
 		if err = ipsMgr.AddToSet(label, podIP, util.IpsetNetHashFlag, podUID); err != nil {
 			metrics.SendErrorLogAndMetric(util.PodID, "[AddPod] Error: failed to add pod to label ipset with err: %v", err)
+			return err
 		}
 	}
 
 	// Add pod's named ports from its ipset.
 	if err = appendNamedPortIpsets(ipsMgr, podContainerPorts, podUID, podIP, false); err != nil {
 		metrics.SendErrorLogAndMetric(util.PodID, "[AddPod] Error: failed to add pod to namespace ipset with err: %v", err)
+		return err
 	}
 
 	// add the Pod info to the podMap
-	npMgr.PodMap[podKey] = npmPodObj
+	c.npMgr.PodMap[podKey] = npmPodObj
 
 	return nil
 }
 
 // UpdatePod handles updating pod ip in its label's ipset.
-func (npMgr *NetworkPolicyManager) UpdatePod(newPodObj *corev1.Pod) error {
-	if !isValidPod(newPodObj) {
-		return nil
-	}
-
-	// today K8s does not allow updating HostNetwork flag for an existing Pod. So NPM can safely
-	// check on the oldPodObj for hostNework value
-	if isHostNetworkPod(newPodObj) {
-		log.Logf(
-			"POD UPDATING ignored for HostNetwork Pod:\n pod: [%s/%s/%s]",
-			newPodObj.ObjectMeta.Namespace, newPodObj.ObjectMeta.Name, newPodObj.Status.PodIP,
-		)
-		return nil
-	}
+func (c *podController) syncAddAndUpdatePod(newPodObj *corev1.Pod) error {
+	podKey, _ := cache.MetaNamespaceKeyFunc(newPodObj)
 
 	var (
 		err            error
-		podKey         = GetPodKey(newPodObj)
 		newPodObjNs    = util.GetNSNameWithPrefix(newPodObj.ObjectMeta.Namespace)
 		newPodObjName  = newPodObj.ObjectMeta.Name
 		newPodObjLabel = newPodObj.ObjectMeta.Labels
 		newPodObjPhase = newPodObj.Status.Phase
 		newPodObjIP    = newPodObj.Status.PodIP
-		ipsMgr         = npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
+		ipsMgr         = c.npMgr.NsMap[util.KubeAllNamespacesFlag].IpsMgr
 	)
 
-	if podKey == "" {
-		err = fmt.Errorf("[UpdatePod] Error: podKey is empty for %s pod in %s with UID %s", newPodObjName, newPodObjNs, string(newPodObj.ObjectMeta.UID))
-		metrics.SendErrorLogAndMetric(util.PodID, err.Error())
-		return err
-	}
-
 	// Add pod namespace if it doesn't exist
-	if _, exists := npMgr.NsMap[newPodObjNs]; !exists {
-		npMgr.NsMap[newPodObjNs], err = newNs(newPodObjNs)
+	if _, exists := c.npMgr.NsMap[newPodObjNs]; !exists {
+		c.npMgr.NsMap[newPodObjNs], err = newNs(newPodObjNs)
 		if err != nil {
 			metrics.SendErrorLogAndMetric(util.PodID, "[UpdatePod] Error: failed to create namespace %s with err: %v", newPodObjNs, err)
 		}
 		log.Logf("Creating set: %v, hashedSet: %v", newPodObjNs, util.GetHashedName(newPodObjNs))
 		if err = ipsMgr.CreateSet(newPodObjNs, append([]string{util.IpsetNetHashFlag})); err != nil {
 			metrics.SendErrorLogAndMetric(util.PodID, "[UpdatePod] Error creating ipset %s with err: %v", newPodObjNs, err)
+			return err
 		}
 	}
 
-	cachedPodObj, exists := npMgr.PodMap[podKey]
+	cachedPodObj, exists := c.npMgr.PodMap[podKey]
```

Comment:
If you are handling the podKey in this PR, you also need to change internal ipset struct to cache IP per PODKey and not PodUID

Reply:
Use podkey in ipset as well.